docs/Routes.md

```diff
@@ -168,7 +171,24 @@ fastify.get('/', options, async function (request, reply) {
   reply.send(processed)
 })
 ```
-**Warning:** If you use `return` and `reply.send` at the same time, the first one that happens takes precedence, the second value will be discarded, a *warn* log will also be emitted if the value is not `undefined`.
+* **Warning:** If you use `return` and `reply.send` at the same time, the first one that happens takes precedence, the second value will be discarded, a *warn* log will also be emitted because you tried to send a response twice.
+* **Warning:** You can't return `undefined`. For more details read [promise-resolution](#promise-resolution).
+
+<a name="promise-resolution"></a>
+### Promise resolution
+
+If you use `async/await` or return a promise inside your handler you should be aware of a special behaviour which is  necessary to support the callback and promise control-flow. You can't resolve a promise with `undefined` it is ignored, a *error* log will also be emitted because you pass the control of the response to the promise which can never be resolved. There are only two simple rules to remember:
+
+1. If you want to use `async/await` or promises but respond a value with `reply.send`:
+    - **Don't** `return` any value.
+    - **Don't** forget to call `reply.send`.
+2. If you want to use `async/await` or promises:
+    - **Don't** use `reply.send`.
+    - **Don't** return `undefined`.
+
+This way is a little bit opinionated but it's the only way to support the most common patterns in the Node.js ecosystem. In spite of so much freedom we highly recommend to go with only one style because error handling should be handled in a consistent way within your application.
```

Comment:
> This way is a little bit opinionated but it's the only way to support the most common patterns in the Node.js ecosystem.

In this way, we can support both *callback-style* and *async-await*, with the minimum trade-off.

___

What about this?

Reply:
Why you don't like the current statement? "minimum tradeoff" is self-evident.

Reply:
I really don't like *"This way is a little bit opinionated"* :P